MCprep_addon/spawner/models.py

```diff
@@ -0,0 +1,1350 @@
+
+
+def updateModelList(context):
+	"""Update the model list"""
+	model_file = bpy.path.abspath(context.scene.model_path)
+	if not os.path.isfile(model_file):
+		print("Invalid model blend file path")
+		context.scene.mcprep_props.model_list.clear()
+		return
+
+	temp_model_list = []  # just to skip duplicates
+	model_list = []
+
+	cache = get_model_cache(context)
+	prefix = "Group/" if hasattr(bpy.data, "groups") else "Collection/"
+	for name in cache.get("groups"):
+		if not name:
+			continue
+		if util.nameGeneralize(name).lower() in temp_model_list:
+			continue
+		if util.nameGeneralize(name).lower() == "Rigidbodyworld".lower():
+			continue
+		description = "Place {x} block".format(x=name)
+		model_list.append((prefix+name, name.title(), description))
+		temp_model_list.append(util.nameGeneralize(name).lower())
+
+	# sort the list alphabetically by name
+	if model_list:
+		_, sorted_blocks = zip(*sorted(zip([block[1].lower()
+			for block in model_list], model_list)))
+	else:
+		sorted_blocks = []
+
+	# now re-populate the UI list
+	context.scene.mcprep_props.model_list.clear()
+	for itm in sorted_blocks:
+		item = context.scene.mcprep_props.model_list.add()
+		item.block = itm[0]
+		item.name = itm[1]
+		item.description = itm[2]
+
+
+class face_struct():
+	"""Structure class for preprocessed faces of a mesh"""
+	def __init__(self, normal_coord, global_coord, local_coord):
+		self.n = normal_coord
+		self.g = global_coord
+		self.l = local_coord
+
+
+# -----------------------------------------------------------------------------
+# Mesh swap functions
+# -----------------------------------------------------------------------------
+
+
+class MCPREP_OT_model_path_reset(bpy.types.Operator):
+	"""Reset the spawn path to the default specified in the addon preferences panel"""
+	bl_idname = "mcprep.model_path_reset"
+	bl_label = "Reset model path"
+	bl_options = {'REGISTER', 'UNDO'}
+
+	@tracking.report_error
+	def execute(self, context):
+
+		addon_prefs = util.get_user_preferences(context)
+		context.scene.model_path = addon_prefs.model_path
+		updateModelList(context)
+		return {'FINISHED'}
+
+
+class MCPREP_OT_model_spawner(bpy.types.Operator):
+	"""Instantly spawn built-in model blocks into a scene"""
+	bl_idname = "mcprep.model_spawner"
+	bl_label = "Model Spawner"
+	bl_options = {'REGISTER', 'UNDO'}
+
+	# properties, will appear in redo-last menu
+	def swap_enum(self, context):
+		return getModelList(context)
+
+	block = bpy.props.EnumProperty(items=swap_enum, name="Model")
+	location = bpy.props.FloatVectorProperty(
+		default=(0,0,0),
+		name = "Location")
+	append_layer = bpy.props.IntProperty(
+		name="Append layer",
+		default=20,
+		min=0,
+		max=20,
+		description="Set the layer for appending groups, 0 means same as active layers")
+	prep_materials = bpy.props.BoolProperty(
+		name="Prep materials",
+		default=True,
+		description="Run prep materials on objects after appending")
+	snapping = bpy.props.EnumProperty(
```

Comment:
Question: does snapping make sense for this operator? Doesn't hurt to keep, but I'm thinking it's less relevant as entities are restricted to being placed like blocks are.

Reply:
I agree, Will remove it